src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs

```diff
@@ -0,0 +1,233 @@
+            }
+
+            // Assumption: pbCipherText := { keyModifier || nonce || encryptedData || authenticationTag }
+            var plaintextBytes = ciphertext.Count - (KEY_MODIFIER_SIZE_IN_BYTES + NONCE_SIZE_IN_BYTES + TAG_SIZE_IN_BYTES);
+            var plaintext = new byte[plaintextBytes];
+
+            try
+            {
+                // Step 1: Extract the key modifier from the payload.
+
+                int keyModifierOffset; // position in ciphertext.Array where key modifier begins
+                int nonceOffset; // position in ciphertext.Array where key modifier ends / nonce begins
+                int encryptedDataOffset; // position in ciphertext.Array where nonce ends / encryptedData begins
+                int tagOffset; // position in ciphertext.Array where encrypted data ends
+
+                checked
+                {
+                    keyModifierOffset = ciphertext.Offset;
+                    nonceOffset = keyModifierOffset + KEY_MODIFIER_SIZE_IN_BYTES;
+                    encryptedDataOffset = nonceOffset + NONCE_SIZE_IN_BYTES;
+                    tagOffset = encryptedDataOffset + plaintextBytes;
+                }
+
+                var keyModifier = new ArraySegment<byte>(ciphertext.Array!, keyModifierOffset, KEY_MODIFIER_SIZE_IN_BYTES);
+
+                // Step 2: Decrypt the KDK and use it to restore the original encryption and MAC keys.
+                // We pin all unencrypted keys to limit their exposure via GC relocation.
+
+                var decryptedKdk = new byte[_keyDerivationKey.Length];
+                var derivedKey = new byte[_derivedkeySizeInBytes];
+
+                fixed (byte* __unused__1 = decryptedKdk)
+                fixed (byte* __unused__2 = derivedKey)
+                {
+                    try
+                    {
+                        _keyDerivationKey.WriteSecretIntoBuffer(new ArraySegment<byte>(decryptedKdk));
+                        ManagedSP800_108_CTR_HMACSHA512.DeriveKeysWithContextHeader(
+                            kdk: decryptedKdk,
+                            label: additionalAuthenticatedData,
+                            contextHeader: _contextHeader,
+                            context: keyModifier,
+                            prfFactory: _kdkPrfFactory,
+                            output: new ArraySegment<byte>(derivedKey));
+
+                        // Perform the decryption operation
+                        unsafe
+                        {
+                            var nonce = new Span<byte>(ciphertext.Array, nonceOffset, NONCE_SIZE_IN_BYTES);
+                            var tag = new Span<byte>(ciphertext.Array, tagOffset, TAG_SIZE_IN_BYTES);
+                            var encrypted = new Span<byte>(ciphertext.Array, encryptedDataOffset, plaintextBytes);
+                            using var aes = new AesGcm(derivedKey);
+                            aes.Decrypt(nonce, encrypted, tag, plaintext);
```

Comment:
This is the decrypt that fails with:  System.Security.Cryptography.CryptographicException : The computed authentication tag did not match the input authentication tag.

I verified that derviedKey, nonce, and tag match the encrypt versions in the CngGcm side